web/modules/custom/brms_sor/src/Service/FmsSorApi.php

```diff
@@ -0,0 +1,252 @@
+      // Throw the exception to the queue processor so the item stays in the
+      // queue and can be retried.
+      throw $e;
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function patch(string $call, array $data = [],): array {
+    try {
+      $token = $this->getAccessToken();
+      $headers = [
+        'Authorization' => 'Bearer ' . $token,
+        'Content-type' => 'application/vnd.api+json',
+      ];
+
+      $args = ['headers' => $headers, 'json' => $data];
+      $uri = $this->baseUrl . $call;
+      $this->response = $this->httpClient->patch($uri, $args);
+      return $this->json->decode($this->response->getBody()->getContents());
+    }
+    catch (RequestException $e) {
+      $this->response = $e->getResponse();
+      $this->logger->get('brms_sor')->error('Failed to call PATCH @uri on FMS API: @message',
+        ['@uri' => $uri, '@message' => $e->getMessage()]);
+      // Throw the exception to the queue processor so the item stays in the
+      // queue and can be retried.
+      throw $e;
+    }
+  }
+
+  /**
+   * Get the UUID of a CFR by its global ID.
+   */
+  protected function getCfrUuid(string $cfr_global_id): ?string {
+    try {
+      // Get the UUID of the CFR with the given global ID in the FMS.
+      $response = $this->get('/asset/cfr', ['filter[cfr_global_id]' => $cfr_global_id]);
+      return $response['data'][0]['id'] ?? NULL;
+    }
+    catch (RequestException $e) {
+      $this->logger->get('brms_sor')->error('Failed to get CFR UUID for global ID @cfr_global_id: @message',
+        ['@cfr_global_id' => $cfr_global_id, '@message' => $e->getMessage()]);
+      // Throw the exception to the queue processor so the item stays in the
+      // queue and can be retried.
+      throw $e;
+    }
+  }
+
+  /**
+   * Update the polygon of a CFR.
+   */
+  public function updatePolygon(string $global_id, ?string $polygon): void {
+    // Get the UUID of the CFR.
+    $uuid = $this->getCfrUuid($global_id);
+    if ($uuid) {
+      // Call the API to update the CFR.
+      $body['data'] = [
+        'type' => 'asset',
+        'id' => $uuid,
+        'attributes' => ['intrinsic_geometry' => $polygon],
+      ];
+      try {
+        $response = $this->patch('/asset/cfr/' . $uuid, $body);
+      }
+      catch (RequestException $e) {
+        $this->logger->get('brms_sor')
+          ->error('Failed to update CFR polygon for global ID @cfr_global_id: @message', [
+            '@cfr_global_id' => $global_id,
+            '@message' => $e->getMessage(),
+          ]);
+        // Throw the exception to the queue processor so the item stays in the
+        // queue and can be retried.
+        throw $e;
+      }
+    }
+    else {
+      $this->logger->get('brms_sor')
+        ->error('Failed to update CFR polygon for global ID @cfr_global_id: CFR not found.', [
+          '@cfr_global_id' => $global_id,
+        ]);
+      // We don't throw an exception here because this is a permanent error and
```

Comment:
Well, I've been here and it might not be fatal, just the other side or the network down for a little bit, but if we're doing a self-correcting system, we can run with this.